modules/pubsub/broker.go

```diff
@@ -0,0 +1,101 @@
+package pubsub
+
+import (
+	"sync"
+
+	"github.com/spaceuptech/space-cloud/model"
+	"github.com/spaceuptech/space-cloud/utils"
+)
+
+natsWildcard string = ">"
+
+type subscription interface {
+	Unsubscribe() error
+}
+
+type pubsubBroker interface {
+	Publish(subject string, msg *model.PubsubMsg) error
+	QueueSubscribe(subject, group string, ch chan *model.PubsubMsg) (*subscription, error)
+}
+
+type pubsubSubscription {
+	subs *subscription
+	ch   chan *model.PubsubMsg
+}
+
+func (m *Module) Publish(subject string, data *model.PubsubMsg) error {
+	return m.connection.Publish(subject, data)
+}
+
+func (m *Module) Subscribe(clientId, subject string, cb model.PubsubCallback) (error) {
+	return m.QueueSubscribe(clientId, subject, clientId, cb)
+}
+
+func (m *Module) QueueSubscribe(clientId, subject, group string, cb model.PubsubCallback) (error) {
+	ch := make(chan *model.PubsubMsg, 10)
+	subject = utils.SingleLeadingTrailing(subject, "/")
+	subs, err := m.connection.QueueSubscribe(subject + natsWildcard, group, ch)
+	if err != nil {
+		return err
+	}
+	go func() {
+		for msg := range ch {
+        	cb(msg.Subject, msg)
+		}
+	}()
+	m.storeSubs(subject, clientId, &pubsubSubscription{subs, ch})
+	return nil
+}
+
+func (m *Module) Unsubscribe(clientId, subject string) err {
+	subject = utils.SingleLeadingTrailing(subject, "/")
+	_subj, ok := m.storage.Load(subject)
+	if ok {
+		subj := subj.(sync.Map)
+		_subs, ok := subj.Load(clientId)
+		if ok {
+			subs := _subs.(pubsubSubscription)
+			subj.Delete(clientId)
+			if mapLen(subj) == 0 {
+				m.storage.Delete(subject)
+			}
+			subj.Range(func (key, value interface{}))
+			subs.subs.Unsubscribe()
+			subs.ch.Close()
+		}
+	}
+	return nil
+}
+
+func (m *Module) UnsubscribeAll(clientId string) err {
+	m.storage.Range(func(subject, v interface{}) bool {
+		m.Unsubscribe(clientId, subject)
+		return true
+	})
+	return nil
+}
+
+//Map subject -> clientId -> pubsubSubscription
+func (m *Module) storeSubs(subject, clientId string, subs *pubsubSubscription) {
+	client, ok := m.storage.Load(subject)
+	if ok {
+		cli := client.(sync.Map)
+		cli.Store(client, subs)
+		return;
+	}
+	var m sync.Map
+	m.Store(client, subs)
+	m.storage.Store(subject, m)
```

Comment:
Seems like you are overwriting all clients here. Instead you must check if a client with that id already exists using LoadOrStore. If it does, throw an error.